Fix multi-selection on value relation widget using string fields

signedav opened the pull request:
In the Value Relation Widget there is the `allowMulti` possibility allowing to store multiple reference values. On using string fields, they are stored on the widget as a formatted string like this: `{1,2,3}`

This fixes #30557

***Why not making conversion on provider side?***
During the improvement of using JSON fields in Postgresql and Geopackages there has been the possibility to pass the values as e.g. lists from the provider to the widget and back, without making any conversation. A provider reads JSON-fields as JSON and stored it back there in the end. The widget can use the `QVariant` (in this case `QVariantMap`) as it is. But in case the field is in the source only a string, the provider does not care if this string contains a list or not. It should not care, because it does not know what this string - however formatted - is used for. That's why it should pass the string as a string to the widget. The value relation widget decodes the formatted string to a list, uses it as a list, and in the end it needs to convert it back into a string, in case that it's not coming from a JSON field.

***Format***
The `{1,2,3}` list format is coming historically - I guess - from HSTORE. In case the provider recognizes an HSTORE (in PG), it possibly could convert it already to a data format and the widgets can use it like that \*. But on a text field, the provider does not have the competence to decide if this string is used as a list or as a string or as whatever on the widget side.

\* this is not the case. Hstore are only maps not lists. I have been missleaded because of old comments calling it hstore...

***GDAL Version***
Some tests are expecting JSON fields in Spatialite Files, what requests GDAL version >= 2.4.
That's why i made the `#ifdef`s there.

***Open Problems***
- [x] The test `testWithJsonInSpatialiteTextFk` will pretty sure fail, since it's not possible to handle key values containing `,` and `{` with the formatted strings. I guess this requests an improvement. -> solved

signedav committed:
store as formatted string if multi-selection on string field

fixes #30557

signedav committed:
test for storing multi-selection in string

signedav committed:
cases with GDAL < 2.4.

signedav committed:
comment

elpaso commented:
@signedav regarding the "open problem" with `testWithJsonInSpatialiteTextFk` I think that it clearly shows all the problems we have with hstore format. JSON is much more robust regarding (un)serialization.

Anyway, when setting the value in the widget we already accept both formats, see `QgsValueRelationFieldFormatter::valueToStringList`

elpaso commented on src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp:
```diff
@@ -75,26 +75,34 @@ QVariant QgsValueRelationWidgetWrapper::value() const
+    else
+    {
+      //store as a formatted string because the fields supports only string
+      v = selection.join( ',' ).prepend( '{' ).append( '}' );
```
Are you sure about this?
Wouldn't be (much) better to use JSON in this case? Providers can convert from JSON to whatever internal format they use, depending on the field type, without loosing information. So, for instance, PG will convert to `hstore` only when needed, and keep JSON in case the field type accepts is.
OTOH, if you stick to `hstore` format, you risk any kind of issue with string keys containing `{},`

signedav replied:
I changed it back to how it has been before. 

I thought about to use JSON for that. But I wonder why it has been in this format at all. Is it just historical or is there a good reason for it? I don't know. Do you? Are there no risks using JSON format instead of HSTORE?

elpaso replied:
I think that `hstore` was the first implementation of this feature, but it is native to PG.

In any event, `hstore` apparently support string keys, so if you want to use `hstore` safely you MUST change the logic that builds and `hstore` by properly quoting the keys (if they are not convertible to `int`), in other words: _never assume that keys are numeric_.

For instance (untested):
```
{ "a {} , \"key" => "a value \" with wierd chars {}," , 1234 => "A value with an int key" }

```

See: https://www.postgresql.org/docs/9.0/hstore.html

**The alternative would be to switch to JSON as the default representation.**

The only potential issue that comes up to my mind is that we might have existing legacy databases that rely on that `hstore` string representation of arrays, that might break if we switch to a JSON representation in case the same values are accessed by another application which is not QGIS (probably not that common use case though).

So, we might consider a transitory logic (until QGIS 4) that tries to guess if the storage format for a string field is `hstore` or `JSON` and act accordingly (issuing a warning in the logs it the format is `hstore`).

If the DB is empty we could just use JSON as a default string representation.

We shouldn't have a problem when reading values from the provider, because we still accept both representations (JSON and hstore).

**Conclusion**

I think that if the quoting of `hstore` keys works fine, we can continue to use `hstore` as a default string representation for arrays without risks. This has the advantage that we won't break any legacy code.

Switching to JSON has the advantage that we would be using a well known, well tested, standardized format with robust conversion libraries, but we would risk the legacy code issue mentioned early (even if I believe the risk is rather low).

signedav replied:
Thanks a lot.

So I see three possibilities:
- improve the `hstore` format (add quotes to text values) like you suggested above
- convert it in value relation widget (if no JSON-field) to a JSON formatted string
- keep it as a `QVariantList` (like before this pr) but add in OGR provider the functionality to convert (if no JSON-field) to a JSON formatted string (in case of the values type is a `QVariantList` and not `QString`)

Because of the - even low - risk of breaking legacy or external code I would keep it in improved hstore format. Maybe we can change that in future...

elpaso replied:
:+1:  sounds like a good plan. 

Would that also fix this string representation issue https://github.com/qgis/QGIS/issues/30565 ?

Maybe a small utility function that can convert the maps/lists to a decent string representation would be required.

signedav replied:
I did the string formatting in the widget still. No utility-function, though I like the idea to have one. But since that integration would include the use of this function everywhere else, and we do not have a final decision how we should layout the representing values, I suggest to have it as a separate integration and keep this PR for the fix of the value relation widget. What don't you think @elpaso ?

elpaso replied:
I think that if we don't make a decision we'll hit this kind of issues again and again, and I alto think that having a separate function with full test coverage encourages other developers to not reinvent the (broken) wheel.

signedav replied:
Okay, thanks. I'll continue on that next week.

m-kuhn replied:
Hello, let me throw in my 2 cents too ;)

# JSON

I would like to use JSON, but for what I see this will break working with existing data sources. Detecting the format is also no option for what I can see, because when adding a new feature, there is no existing data that could be checked for JSON vs. hstore. So to continue working and keep things stable for our users we *must* continue using hstore for the lifetime of QGIS 3 since I cannot see an automatic way for switching this to JSON transparently. In short, I don't think the risk of hitting legacy issues is low, on the contrary, I would suspect it is almost guaranteed.

Users who want to use something reasonable will be able to setup a data source using native JSON types (which should be working already and is not affected by this pull request)

If we want e.g. shapefile users to use JSON too for this, this should be treated as a - very welcome - "feature request" (new option for the relation widget) and not as a "bug fix".

# Serializing / parsing

If it helps to put hstore serializing/parsing into a public QGIS function, no objections from my side :+1:

signedav replied:
Thanks.
Yes, I'll make an hstore utility function. 
Any inputs where I should place it?

Gustry replied:
> Yes, I'll make an hstore utility function. Any inputs where I should place it?

https://github.com/qgis/QGIS/blob/master/src/core/qgshstoreutils.h ?

elpaso replied:
> Hello, let me throw in my 2 cents too ;)
> # JSON
> 
> I would like to use JSON, but for what I see this will break working with existing data sources. Detecting the format is also no option for what I can see, because when adding a new feature, there is no existing data that could be checked for JSON vs. hstore. So to continue working and keep things stable for our users we _must_ continue using hstore for the lifetime of QGIS 3 since I cannot see an automatic way for switching this to JSON transparently. In short, I don't think the risk of hitting legacy issues is low, on the contrary, I would suspect it is almost guaranteed.
> 

Well, I partially agree, I think that we have two scenarios here: either we have a new table (or a table with no data in the "map" column) or we have some data in it. If we have some data we could sniff the format and act accordingly.

> Users who want to use something reasonable will be able to setup a data source using native JSON types (which should be working already and is not affected by this pull request)
> 
> If we want e.g. shapefile users to use JSON too for this, this should be treated as a - very welcome - "feature request" (new option for the relation widget) and not as a "bug fix".
> # Serializing / parsing
> 
> If it helps to put hstore serializing/parsing into a public QGIS function, no objections from my side +1
> # Summary
> 
> Let's make sure this pull request fixes things and does not introduce any other regression.

Of course. But I don't get the point: what I'm trying to say here it that this undecided behavior about the internal and the UX (presentation) of complex (map and array values) has been an endless source of issues and apparently when you fix it in one side you break it in another place, it is this attitude to patch it without addressing the real problem than constantly led to this situation (I'm responsible as well, mea culpa).

Additionaly, the UX representation was naive at it best: you cannot just split and join on a comma to get it right.

I do not insist at this point, let's live with it but at least let's find a robust way to represent the values to the user.

m-kuhn replied:
I have been trying to improve the situation by introducing native types (jsonb etc. for pg, json as data type for gpkg). I think that's fairly stable and I can't see an endless source of issues there.

On the other hand this pull request here is about a regression that plain hstore (pg array) formatted values no longer worked. My only point was that this should be the scope of this PR and that we should eventually be able to work with QGIS 3.10 and 3.4 on the same data without interferences :)

I'm happy if we have the code (possibly with the same deficiencies) working again. If there's a clear path to fix *additional* things, that's good. If it requires more (like cross-checking how exactly postgres handles it etc.) this can easily come in a followup step and should not block this fix of a regression.

signedav replied:
Regarding the utility function to convert the list in the hstore-like `{1,2,3}` formatted string, I think `qgshstoreutils.h` is not the right place for it. 

Since the naming "hstore" is not very good, because it's not `hstore` - it's how Postgres represents an array. So I would prefer to create a `QgsPgArrayUtils` containing functions to convert a list to this String format and back and try to avoid the naming `hstore` for these formatted strings wherever possible...

elpaso replied:
I agree, the term `hstore` has played a substantial role in muddying the waters.

troopa81 replied:
> I would prefer to create a QgsPgArrayUtils

@signedav That's pretty much what I propose [here](https://github.com/qgis/QGIS/pull/9610#issuecomment-510065660) (I try to fix a quite similar issue). I will prefer a non specific Postgres solution. 

We can extract code to parse array from postgres provider (in a new QgsArrayUtils class?). `getNextString` method can already deal with a given separator. It will be easy to improve if some day we want to deal with different format (different brackets, different separator).

elpaso commented on tests/src/gui/testqgsvaluerelationwidgetwrapper.cpp:
```diff
@@ -624,9 +646,181 @@ void TestQgsValueRelationWidgetWrapper::testWithJsonInGPKG()
+  // create ogr gpkg layers
+  QString myFileName( TEST_DATA_DIR ); //defined in CmakeLists.txt
+  QString myTempDirName = tempDir.path();
+  QFile::copy( myFileName + "/provider/test_json.gpkg", myTempDirName + "/test_json.gpkg" );
```
I think that for extra safety you may add test cases with `{},"'` in author names to this test file.

Also, if isn't there a PG test yet, we should prepare a table with string PK and  `{},"'` in the PK values and check it with both JSON and hstore field types.

signedav committed:
use quotes in hstore list items

signedav committed:
testcase wiht string keys

signedav committed:
placed fallback better

signedav committed:
improve tests with wrong hstore list

signedav committed:
finally right placed fallback

elpaso commented on src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp:
```diff
@@ -180,7 +180,14 @@ QStringList QgsValueRelationFieldFormatter::valueToStringList( const QVariant &v
     {
       newVal = QVariant( newVal.toString().trimmed().mid( 1 ).mid( 0, newVal.toString().length() - 2 ).prepend( '[' ).append( ']' ) );
+
+      if ( !json::accept( newVal.toString().toStdString() ) )
```
what accept does? 
If you are trying to parse json, you probably want `parse()` within a try/catch block https://nlohmann.github.io/json/classnlohmann_1_1basic__json_afd4ef1ac8ad50a5894a9afebca69140a.html#afd4ef1ac8ad50a5894a9afebca69140a

signedav replied:
It's to check if the replacement of the `{}` resulted in a correct JSON format (means that the `parse` can be successful later). If not (because the data are still stored earlier without quotes), it decodes the hstore-list like in the old days.

elpaso replied:
yeah, I guessed that it was your intention, the usage seems correct, I was mislead by another `accept`: https://nlohmann.github.io/json/classnlohmann_1_1basic__json_ae797958b922732bf5fc01053d0659c1f.html#ae797958b922732bf5fc01053d0659c1f

signedav committed:
escape backslash to escape quote

signedav committed:
test for weird characters in key strings for text stored values (in hstore format)

signedav committed:
spelling correction

signedav committed:
escape backslashes as well

elpaso commented on src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp:
```diff
@@ -75,26 +75,53 @@ QVariant QgsValueRelationWidgetWrapper::value() const
+            break;
+          default:
+            QString newS = s;
+            newS.replace( '\\', QStringLiteral( "\\\\" ) );
```
Please use raw string here, much more readable than counting slashes.
```suggestion
            newS.replace( '\\', QStringLiteral( R"(\\)" ) );
```

m-kuhn commented on src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp:
```diff
@@ -75,26 +75,53 @@ QVariant QgsValueRelationWidgetWrapper::value() const
-    QVariantList vl;
-    //store as QVariantList because it's json
-    for ( const QString &s : qgis::as_const( selection ) )
+    if ( layer()->fields().at( fieldIdx() ).type() == QVariant::Map )
```
Does this need to check for list too?

signedav replied:
Not sure by heart. I thought it has always the type map. But I will doublecheck.

signedav replied:
Well apparently the content  is handled as JSON (means `QVariant::Map` or as this hstore-like list-format `{1,2,3}}` (means: `QVariant::String`) so we do not need to check for list. Or am I'm wrong?

m-kuhn replied:
```
git grep -n QVariant::List
arcgisrest/qgsarcgisrestutils.cpp:152:    if ( curveData.type() == QVariant::List )
oracle/ocispatial/qsql_ocispatial.cpp:1850:    fieldTypes.append( tp == QVariant::List ? boundValues.at( i ).toList().value( 0 ).type()
oracle/ocispatial/qsql_ocispatial.cpp:1863:    if ( boundValues.at( i ).type() != QVariant::List )
oracle/ocispatial/qsql_ocispatial.cpp:2138:    if ( tp != QVariant::List )
postgres/qgspostgresconn.cpp:1053:    case QVariant::List:
postgres/qgspostgresprovider.cpp:238:      << QgsVectorDataProvider::NativeType( tr( "Array of number (integer - 32bit)" ), QStringLiteral( "int4[]" ), QVariant::List, -1, -1, -1, -1, QVariant::Int )
postgres/qgspostgresprovider.cpp:239:      << QgsVectorDataProvider::NativeType( tr( "Array of number (integer - 64bit)" ), QStringLiteral( "int8[]" ), QVariant::List, -1, -1, -1, -1, QVariant::LongLong )
postgres/qgspostgresprovider.cpp:240:      << QgsVectorDataProvider::NativeType( tr( "Array of number (double)" ), QStringLiteral( "double precision[]" ), QVariant::List, -1, -1, -1, -1, QVariant::Double )
postgres/qgspostgresprovider.cpp:1053:        fieldType = ( fieldType == QVariant::String ? QVariant::StringList : QVariant::List );
postgres/qgspostgresprovider.cpp:3838:    case QVariant::List:
postgres/qgspostgresprovider.cpp:4459:  if ( ( type == QVariant::StringList || type == QVariant::List ) && inner.startsWith( "{" ) )
postgres/qgspostgresprovider.cpp:4479:    case QVariant::List:
spatialite/qgsspatialitefeatureiterator.cpp:580:    if ( type == QVariant::List || type == QVariant::StringList )
spatialite/qgsspatialiteprovider.cpp:95:    case QVariant::List:
spatialite/qgsspatialiteprovider.cpp:605:                  << QgsVectorDataProvider::NativeType( tr( "Array of decimal numbers (double)" ), SPATIALITE_ARRAY_PREFIX.toUpper() + "REAL" + SPATIALITE_ARRAY_SUFFIX.toUpper(), QVariant::List, 0, 0, 0, 0, QVariant::Double )
spatialite/qgsspatialiteprovider.cpp:606:                  << QgsVectorDataProvider::NativeType( tr( "Array of whole numbers (integer)" ), SPATIALITE_ARRAY_PREFIX.toUpper() + "INTEGER" + SPATIALITE_ARRAY_SUFFIX.toUpper(), QVariant::List, 0, 0, 0, 0, QVariant::LongLong )
spatialite/qgsspatialiteprovider.cpp:664:    return TypeSubType( subType.first == QVariant::String ? QVariant::StringList : QVariant::List, subType.first );
spatialite/qgsspatialiteprovider.cpp:668:    return TypeSubType( QVariant::List, QVariant::Invalid );
spatialite/qgsspatialiteprovider.cpp:4097:          else if ( type == QVariant::StringList || type == QVariant::List )
spatialite/qgsspatialiteprovider.cpp:4404:        else if ( type == QVariant::StringList || type == QVariant::List )
```

I guess the one where spatialite provides arrays as lists could potentially be interesting for us here?

signedav replied:
Hm... okay, havent seen that. We can consider that and add the check for list.

signedav committed:
qgsarrayutils to care for postgres-array like formatted list in strings

signedav committed:
sip stuff

signedav committed:
sip in file

signedav committed:
tests for arrayutils creating lists and parsing back

m-kuhn commented on src/core/qgsarrayutils.cpp:
```diff
@@ -0,0 +1,94 @@
+#include <nlohmann/json.hpp>
+using json = nlohmann::json;
+
+QVariantList QgsArrayUtils::parse( const QString &string )
```
Should this be called `PgArrayUtils` since it's a very particular array style? (And the only other array style in use is already handled by the `JsonUtils`, so making it highly configurable is of limited use at the moment for what I can see).

signedav replied:
It's this general name in case we want to extend the utils to other formats. But I can rename it to `PgArrayUtils` if you want... I'm not fixed on that.

m-kuhn replied:
Right now it's called `QgsArrayUtils::parse`, so it sounds like it's generic but it's implemented specific. I think either the method name or the class name should reflect the specifity.

signedav replied:
I agree. I vote for the method name...

signedav replied:
After thinking about. In case we replace all the postgres array parse functions with this it's maybe better to call it qgsPgArrayUtils...

troopa81 replied:
> After thinking about. In case we replace all the postgres array parse functions with this it's maybe better to call it qgsPgArrayUtils...

Is there any reason for not using the actual postgres provider code? Do you plan to support the multidimensional array or provider specific type (like a boolean array)?

signedav replied:
In case there is a function parsing and building postgres arrays it would make sense when the postgres provider uses them too, right? But when the postgres provider uses functions there, it would probably make sense to move all parsing/building functions there too. But I'm not sure if I want to make this massive change. I'm checking out what would have to be done then...

troopa81 replied:
I'm quite sure that both postgresprovider and widget that need to parse array needs to call the same code to parsearray. 

Possibly, the one you are currently proposing. But there is already a bunch of methods in postgres provider (getNextString, parseArray, parseStringArray...) that already parse array and deals with some issues you are facing to (quotes, escaped character, multidimensionnal array...). 

Why not copying them in a QgsArrayUtils class instead of rewriting a new parser? But maybe I miss something.

signedav replied:
Copying them is anyway a possibility I am heading to. But maybe in the future we can reduce it to one function used everywhere.

m-kuhn commented on src/core/qgsarrayutils.cpp:
```diff
@@ -0,0 +1,94 @@
+  {
+    newVal =  newVal.trimmed().mid( 1 ).mid( 0, newVal.length() - 2 ).prepend( '[' ).append( ']' );
+
+    if ( !json::accept( newVal.toStdString() ) )
```
Is this what postgres does? I.e. is the postgres quoting arrays the same style as JSON? And nested arrays and maps are expected in JSON structure?

signedav replied:
I adapted this way of work from the original `QgsValueRelationWidgetFormatter` function to decode it as a first step. I did not replace the `QgsPostgresConn::quotedList` and `QgsPostgresProvider::parseArray``yet.

m-kuhn replied:
I see. This work was introduced at the same time as the de(serialization) code in this widget was removed, so it's quite young too.

Authoritative here should be how postgres handles this (and not how our postgres provider implements it or how QGIS 3.8 thinks it should be done). Let me have a look ;)

m-kuhn replied:
Some random tests on postgres

| string                          | result                       |
|---------------------------------|------------------------------|
| SELECT '{1, 2}'::int[]          | [1, 2]                       |
| SELECT '{a, b}'::text[]         | ['a', 'b']                   |
| SELECT '{"a", "b"}'::text[]     | ['a', 'b']                   |
| SELECT '{ "a\"", "b" }'::text[] | ['a"', 'b']                  |
| SELECT '{a [], b}'::text[]      | ['a []', 'b']                |
| SELECT '{a {}, b}'::text[]      | UB / Malformed array literal |
| SELECT '{a \, b}'::text[]       | ['a , b']                    |

signedav replied:
In the end it's: write strings in quotes and escape doublequotes in a strings.

signedav commented on src/core/qgsarrayutils.cpp:
```diff
@@ -0,0 +1,94 @@
+  QString newVal = string;
+  if ( newVal.trimmed().startsWith( '{' ) )
+  {
+    newVal =  newVal.trimmed().mid( 1 ).mid( 0, newVal.length() - 2 ).prepend( '[' ).append( ']' );
```
You are right. Actually it cannot decode it with JSON because the "subarrays" are not quoted. I see what I can do. In the `qgspostgresprovider` we support multidimensional array,  but as far I understand it, we create still a one dimensional list containing "subarrays" as string, right?

troopa81 replied:
Yes, right. It could manage a real multidimensional array by calling recursively parseMultiDimensionnalArray but it's useless because there is no widget that know what to do with such data.

signedav committed:
improved parsing and building function to handle multidimensional arrays
code mostly taken from this integration in the postgresprovider / postgresconn
x

signedav committed:
renamed QgsArrayUtils to QgsPostgresStringUtils

signedav committed:
remove unused sip.in

signedav commented:
First, in my opinion it would make sence to move the parsing and building functions from the `QgsPostgresProvider` / `QgsPostgresConn` to a general utility like this one, but after some research I've seen that these are quite a lot. This implementation here is actually a Bugfix and I would suggest to keep the moving of all the functions to a follow up.

Second, I like @elpaso s integration with the conversion to JSON array and then parse with JSON but I cared about the request of @troopa81 beeing able to parse and build multidimensional arrays. There the issue is, that an element in an array could look like `{1,2,3}` without any leading quotes. Only way would have been to replace all the `{}` with `[]` but this would change the content of the list, because as it's implemented in `QgsPostgresProvider` it stores the subarrays as strings as they are (means wit `{}`). So I took over the way of work of `QgsPostgresProvider` and `QgsPostgresConn`.

Third, I renamed it now to `QgsPostgresStringUtils` to be open to integrate the Postgres parse/build functionality there. Actually it will be always with string in the end.

m-kuhn commented:
Nice one. And good job with all the tests!

Only some small things here from my side:

- Rename build and parse to buildArray and parseArray
- I'd make the `getNextString` method private (implementation detail)
- Check the (test) file headers
- Check the [indentation here](https://github.com/qgis/QGIS/pull/30610/files#diff-81459b608a9e0a81d0e5cc3965566c13R1559)

If @troopa81 and @elpaso are happy with it too, let's merge.

signedav committed:
made QgsPostgresStringUtils namespace to class because of private functions we use there and possibly in future there are more coming and renamed the methods
fixed indents and comments

troopa81 commented:
Thanks for the work, I'm expecting this PR to be merged so I can finally fix mine https://github.com/qgis/QGIS/pull/9610 :)

I'm a little bit concern about few things:
- QgsPostGresProvider don't use this class to parse array, so there is duplicate code meaning duplicate maintenance
- There is no call to a `convertValue` method like when parsing PG like array so you end up with a string array instead of a int array for instance. It prevent reusing it from postgres provider or from listwidget. 
- The class name is PostGresStringUtils but we parse also array in non PG-format []. And in the other side, we can't buildArray with []

troopa81 commented:
> to put all of them to these utils goes beyond the scope of this bugfix IMO

I can live with that. IMO, It will have to be done and I hope the two code won't change too much in the meantime. 

> Should it convert a value in the array to an int if it's numeric, in your opinion? 

Indeed, convertion could be done later in the process, so I'm ok with they way it is.  

> I keep that in case someone had this style of list in a non-json field

You mean a PG array that would be described with [] instead of {} ? It's not suppose to happen because PG array are always defined with {}. And if you detect you have a json array, it seems better to not use QgsPostgresStringUtils but a json specific code (QgsJSONUtils for instance).

signedav commented:
> You mean a PG array that would be described with [] instead of {} ? It's not suppose to happen because PG array are always defined with {}. And if you detect you have a json array, it seems better to not use QgsPostgresStringUtils but a json specific code (QgsJSONUtils for instance).

For the Value Relation Widget data source fields with the type string/text can be used, so there can be stored every style. Question is, how the widget decodes it and encodes it again in the end. It has always been the PG Array format and will be in future, but on this implementation https://github.com/qgis/QGIS/commit/564fd50e433f7fcee66fabe5d569da470c5b034d the possibility popped up to have a source with string-values as JSON arrays.
I think this decoding snipped exists only to encode the string after having replaced the `{}` with `[]` and no one ever used it to decode his JSON formatted strings, but I'm not sure and so I keeped it. But if @elpaso agrees, I will remove that part at https://github.com/qgis/QGIS/pull/30610/files#diff-b0d816c67a40dad0c634ac7a65484c53R116

troopa81 commented:
> but on this implementation
> https://github.com/qgis/QGIS/commit/564fd50e433f7fcee66fabe5d569da470c5b034d
> the possibility popped up to have a source with string-values as JSON
> arrays

If I understand well the type of array depends on the provider (
Spatialite -> JSON array, PostGres -> HStore array). That was one of the problem I was
trying to fix with PR #9610. And we decide to add an enum format in QgsField so the
provider can specify which format he wants to use.

I propose to let the JSON parsing in
`QgsValueRelationFieldFormatter::valueToStringList` method, if it starts
with {, we call the QgsPostgresStringUtils methods else we use the json
parsing code.

I will change the if condition to base it on the new format enum field
when finishing my PR.

signedav commented:
> If I understand well the type of array depends on the provider (
Spatialite -> JSON array, PostGres -> HStore array).

That's not what I understand. IMO it can be JSON as well in PostGres when it's a `json/jsonb` field. And with Spatialite it's only JSON as type, in case the Spatialite field has the type JSON (as subtype of STRING). But then it's decoded by the provider anyway and passes a list or a map to the widget. 
But in this case it's on provider side only a text. Nothing else. But in this text it's written "{1,2,3}". And that's why the Widget calls this function to convert this simple text with the type string to a data type it can work with (list).

> I propose to let the JSON parsing in
`QgsValueRelationFieldFormatter::valueToStringList` method, if it starts
with {, we call the QgsPostgresStringUtils methods else we use the json
parsing code.

Fine for me.

elpaso commented:
> but on this implementation [564fd50](https://github.com/qgis/QGIS/commit/564fd50e433f7fcee66fabe5d569da470c5b034d) the possibility popped up to have a source with string-values as JSON arrays
> If I understand well the type of array depends on the provider ( Spatialite -> JSON array, PostGres -> HStore array). That was one of the problem I was trying to fix with PR #9610. And we decide to add an enum format in QgsField so the provider can specify which format he wants to use. I propose to let the JSON parsing in `QgsValueRelationFieldFormatter::valueToStringList` method, if it starts with {, we call the QgsPostgresStringUtils methods else we use the json parsing code. I will change the if condition to base it on the new format enum field when finishing my PR.

Actually, the root of the problem is that the way providers store "complex" values (arrays and dictionaries) can vary, for instance, PG can handle JSON(B) fields, HSTORE and plain varchars and the widget should be tolerant enough to handle all these cases.

So, my first though (discussed with @m-kuhn ) was that we should delegate the conversion to-from QVariant(List/Map) to the providers, so that the widgets do not need to know what is the final provider-specific encoding format for the values and we can treat complex values as QVariantList/Map through the whole QGIS code base (widgets and formatters etc.) without worrying about the actual storage encoding.

However, the problem is that in case the values are stored as strings (varchars or text), the provider does not know whether the string contains "complex" encoded values and in which encoding they are stored (hstore, json or the fancy future encoding format that will be invented tomorrow), only the widget knows that.

So, we are forced to stick with this complex widget logic that tries to guess whether the data are encoded  strings of complex values and in what format (hstore, json etc.) they are stored.

The current strategy (@m-kuhn correct me if I'm wrong) is that we convert the values to QVariant(List/Map) in the provider if the provider knows that the field is a complex storage (JSON, hstore) and pass them to the widget as QVariantList/Map, but we leave them as QVariant::String in case the storage is a string (varchar, text, whatever), in that case the encoding may vary but it will probably be `hstore` or `json` and we should be able to tell one from the other by trying to decode as JSON (because `hstore` will not convert to JSON - I'm not 100% sure here).

This approach is fragile at best, and I would really like to see a common Qgs* utility function that does this guessing and conversion, with full unit tests.

I probably made it worse when I introduced the spatialite json handling, but PG also offers JSON and I'm a big fan of standard encondings that would work across providers rather than a PG specific `hstore` implementation.

signedav committed:
moved fallback json string decode part to widget to have the "PostgresString" function without JSON

signedav commented:
> The current strategy (@m-kuhn correct me if I'm wrong) is that we convert the values to QVariant(List/Map) in the provider if the provider knows that the field is a complex storage (JSON, hstore) and pass them to the widget as QVariantList/Map, but we leave them as QVariant::String in case the storage is a string (varchar, text, whatever), in that case the encoding may vary but it will probably be hstore or json and we should be able to tell one from the other by trying to decode as JSON (because hstore will not convert to JSON - I'm not 100% sure here).

You are right. 
But the Value Relation Widget before did not work with JSON content in a varchar fields. That's what has been introduced, when you improved the decoding in the mentioned commit. But in the end it stores it back in the Postgres-Array format (now and before the mentioned commit) - no matter what it decoded from the varchar originally.

> This approach is fragile at best, and I would really like to see a common Qgs* utility function that does this guessing and conversion, with full unit tests.

So you would prefer to move the "guessing" back to `QgsPostgresStringUtils`? I don't care where it is, but @troopa81 does :-)
And the test case you request is to load it with a JSON content in the string field? Otherwise the tests are fine in your opinion?

elpaso commented:
> > The current strategy (@m-kuhn correct me if I'm wrong) is that we convert the values to QVariant(List/Map) in the provider if the provider knows that the field is a complex storage (JSON, hstore) and pass them to the widget as QVariantList/Map, but we leave them as QVariant::String in case the storage is a string (varchar, text, whatever), in that case the encoding may vary but it will probably be hstore or json and we should be able to tell one from the other by trying to decode as JSON (because hstore will not convert to JSON - I'm not 100% sure here).
> 
> You are right.
> But the Value Relation Widget before did not work with JSON content in a varchar fields. That's what has been introduced, when you improved the decoding in the mentioned commit. But in the end it stores it back in the Postgres-Array format (now and before the mentioned commit) - no matter what it decoded from the varchar originally.
> 

No, I think that the issue I fixed was that we could not handle JSON values stored in JSON fields from spatialite (and PG, but that was not the bug I fixed).

I might be wrong here, but I think that this test https://github.com/qgis/QGIS/commit/564fd50e433f7fcee66fabe5d569da470c5b034d#diff-81459b608a9e0a81d0e5cc3965566c13R629 covers the conversion back to JSON in case of spatialite JSON fields.

> > This approach is fragile at best, and I would really like to see a common Qgs* utility function that does this guessing and conversion, with full unit tests.
> 
> So you would prefer to move the "guessing" back to `QgsPostgresStringUtils`? I don't care where it is, but @troopa81 does :-)

@signedav I don't really care where the method is as long as it works :smile: 

> And the test case you request is to load it with a JSON content in the string field? Otherwise the tests are fine in your opinion?

That would be the final missing piece, we are 100% safe if we can handle (in and out from the widget)
- QVariantList/Map coming from JSON and hstore native fields from providers that support them (PG, GPKG, spatialite AFAIK)
- hstore encoded strings coming from string/text fields from any provider
- JSON encoded strings coming from string/text fields from any provider

the problem we are trying to solve is always that in case of string/text fields we do not know the encoding and we need to handle that inside the widget and pass it back to the provider already encoded (hstore or json).

Forgive me if I keep repeating myself, I'm just trying to summarize for my own mental health ;)

signedav commented:
> No, I think that the issue I fixed was that we could not handle JSON values stored in JSON fields from spatialite. 

Okay. But what I meant was, that it has not been possible to decode textfields containing arrays in JSON format before the mentioned commit. If we request that (with in and **out** from the widget) now like ... 
> JSON encoded strings coming from string/text fields from any provider

... the "out"-part is an additional feature containing the part with remembering the format inside widget. If this is a mandatory part of this bug fix for you to agree on this PR, then I can add it. But in my opinion it's - although it's nice - optional.

elpaso commented:
> > No, I think that the issue I fixed was that we could not handle JSON values stored in JSON fields from spatialite.
> 
> Okay. But what I meant was, that it has not been possible to decode textfields containing arrays in JSON format before the mentioned commit. If we request that (with in and **out** from the widget) now like ...
> 
> > JSON encoded strings coming from string/text fields from any provider
> 
> ... the "out"-part is an additional feature containing the part with remembering the format inside widget. If this is a mandatory part of this bug fix for you to agree on this PR, then I can add it. But in my opinion it's - although it's nice - optional.

I agree that it's a corner case: the main use case we want to handle is that a JSON native field is usable inside the widget and I think that that case is already handled (by my original commit).

What I'm afraid of, if you don't implement the "out" conversion correctly, is that a `JSON` encoded string field is silently converted to `hstore` encoding, is that what you want?

signedav commented:
> the main use case we want to handle is that a JSON native field is usable inside the widget and I think that that case is already handled (by my original commit).

Yes. This PR is about to use text fields. Has nothing to do with the JSON native field. 

> What I'm afraid of, if you don't implement the "out" conversion correctly, is that a JSON encoded string field is silently converted to hstore encoding, is that what you want?

Yes. And the reason is this (talking about fields of string/text/varchar type):
Before your commit, it only supported `{1,2,3}`-like formatted lists in the string fields and JSON only in the json fields. After your commit, it supported to read `{1,2,3}`-like formatted lists *and* json formatted lists in the text fields, but as far that I tested that, it could *not store* them back (neither JSON nor `{1,2,3}`-like). That's why I thought now: To fix it, I change it back, like it has been before (means: `{1,2,3}`-like and without JSON). But then I thought: It's rarely, but still possible, that someone filled up his data (from outside or with textedit-widget) with JSON-list and has been able to display it as Value Relation Widget, without saving something. And that's the fallback I still wanted to support. In case now the user stores the data, they are stored in the `{1,2,3}`-like format (instead of loosing the data because writing an empty string). 

So since this has never (I guess) been working properly, we have to ask ourselves the question: Do we want to support in a proper way JSON formatted lists in the text field (as an additional functionality) or should we just go back to the working status (what this PR does until now)?

Btw. if we want to have JSON and `{1,2,3}`-like formats, what should we do when we have an empty field? Store it per default in the old way `{1,2,3}`? But then we could have a data set with mixed formats in the the entries.

elpaso commented:
> So since this has never (I guess) been working properly, we have to ask ourselves the question: Do we want to support in a proper way JSON formatted lists in the text field (as an additional functionality) or should we just go back to the working status (what this PR does until now)?
> 
> Btw. if we want to have JSON and `{1,2,3}`-like formats, what should we do when we have an empty field? Store it per default in the old way `{1,2,3}`? But then we could have a data set with mixed formats in the the entries.

Well, in theory, you could sniff the format from other existing field values (if the table is not empty), but I agree that for now we can use your proposed implementation, go ahead!

troopa81 commented:
@signedav Thanks for the corrections!  it looks fine to me now

m-kuhn commented on src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp:
```diff
@@ -94,7 +95,16 @@ QVariant QgsValueRelationWidgetWrapper::value() const
     }
-    v = vl;
+
+    if ( layer()->fields().at( fieldIdx() ).type() == QVariant::Map )
```
or `QVariant::List`?

signedav committed:
handle variant as list (instead of string) when the field.type() is QVariant::List - since these values are not stored as string, the tests don't need to check the GDAL version (what shouldn't have been done before anyway because it only concerns geopackages, where this version is checked on storing data...)

m-kuhn commented on src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp:
```diff
@@ -96,8 +96,10 @@ QVariant QgsValueRelationWidgetWrapper::value() const
+    if ( layer()->fields().at( fieldIdx() ).type() == QVariant::Map ||
+         layer()->fields().at( fieldIdx() ).type() == QVariant::List )
     {
+      //because of json it's stored as QVariantList with type QVariant::Map or QVariant::List
```
Maybe JSON, maybe some other way for structured data in the data provider. I think that's the wrong level to talk about storage types.

signedav replied:
Okay, i'll remove these comments... In the end it's rather confusing than helpful.

signedav committed:
no confusing comments

signedav commented:
Thanks for your inputs and support @elpaso @troopa81 @m-kuhn

RudiU commented:
I wonder if this solution has already been implemented in version 3.8.1? 
In version 3.8.1 we have the problem with value relation (multi selection) values, when we try to store them in a shapefile (i.e. dBase). While this worked (and still works) with QGIS 3.6, it ends up with empty brackets in QGIS 3.8.1

gioman commented:
> I wonder if this solution has already been implemented in version 3.8.1?
> In version 3.8.1 we have the problem with value relation (multi selection) values, when we try to store them in a shapefile (i.e. dBase). While this worked (and still works) with QGIS 3.6, it ends up with empty brackets in QGIS 3.8.1

I'm not sure this has been backported (can't see the backport labels here).

signedav commented:
You are right. In my opinion it's a bugfix that should be backported. What do the others think?
Should I do it manually, or can it still be done by labels?
